### PR TITLE
Allow querying InstalledVersions::isInstalled() for no-dev packages

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Will help to fix issues like https://github.com/symfony/symfony/issues/40136

### CODE REVIEW

File: tests/Composer/Test/InstalledVersionsTest.php [hunk 1]
```diff
@@ -49,9 +49,10 @@ public function testGetInstalledPackages()
     /**
      * @dataProvider isInstalledProvider
      */
-    public function testIsInstalled($expected, $name, $constraint = null)
+    public function testIsInstalled($expected, $name, $noDev = false)
     {
-        $this->assertSame($expected, InstalledVersions::isInstalled($name));
+        // @phpstan-ignore-next-line
```
Comment (nicolas-grekas) [inline tests/Composer/Test/InstalledVersionsTest.php:54]: Sames from `Composer\InstalledVersions::isInstalled() invoked with 2 parameters, 1 required`

Comment (Seldaek) [reply tests/Composer/Test/InstalledVersionsTest.php:54]: That'll fix itself once the code is merged tho, so I'd rather have useless warnings in the PR than have to remove ignore stuff later :)

File: tests/bootstrap.php [hunk 2]
```diff
@@ -17,4 +17,5 @@
 }
 
 require __DIR__.'/../src/bootstrap.php';
+require __DIR__.'/../src/Composer/InstalledVersions.php';
```
Comment (Seldaek) [inline tests/bootstrap.php:20]: Why this change?

Comment (stof) [reply tests/bootstrap.php:20]: I think this is to enforce that tests run against the code of the source rather than against the code generated in `vendor/` by the phar used to install deps, to be sure that tests use the uptodate code.

Comment (nicolas-grekas) [reply tests/bootstrap.php:20]: @stof is right :)

File: tests/Composer/Test/InstalledVersionsTest.php (hunk 1 above)
Comment (nicolas-grekas) [reply tests/Composer/Test/InstalledVersionsTest.php:54]: removed

File: src/Composer/Repository/FilesystemRepository.php [hunk 3]
```diff
@@ -151,6 +152,7 @@ public function write($devMode, InstallationManager $installationManager)
                     'version' => $package->getVersion(),
                     'aliases' => array(),
                     'reference' => $reference,
+                    'dev' => isset($devPackages[$package->getName()]),
```
Comment (naderman) [inline src/Composer/Repository/FilesystemRepository.php:155]: Wonder if this should be 'require-dev' or something that makes it clear this isn't about a dev version, e.g. 2.3.x-dev but rather about whether the package was installed because of require-dev?

File: src/Composer/InstalledVersions.php [hunk 4]
```diff
@@ -53,13 +53,14 @@ public static function getInstalledPackages()
      * This also returns true if the package name is provided or replaced by another package
      *
      * @param  string $packageName
+     * @param  bool   $noDev
```
Comment (naderman) [inline src/Composer/InstalledVersions.php:56]: Similarly to my other comment, I think a better name would make it clear that this is about dependencies resulting from requireDev rather than dependencies which happen to have a dev version like 2.3.x-dev

File: src/Composer/Repository/FilesystemRepository.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Composer/Repository/FilesystemRepository.php]: The code uses the term "dev-mode", so I adopted it. WDYTN

File: src/Composer/InstalledVersions.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Composer/InstalledVersions.php]: I propose `$noDevMode` then :)

Comment (naderman) [reply src/Composer/InstalledVersions.php]: Works for me

File: src/Composer/Repository/FilesystemRepository.php (hunk 3 above)
Comment (Seldaek) [reply src/Composer/Repository/FilesystemRepository.php]: Sorry for being late to the party but I wonder.. Wouldn't `$noDevRequirements`/`dev-requirement` be more self-explanatory? require-dev => dev requirements.

Comment (naderman) [reply src/Composer/Repository/FilesystemRepository.php]: Yeah I think that works even better, as "dev-mode" may be misunderstood as simply whether composer was run with --dev or not. Yay bikeshedding, but let's go with @Seldaek's suggestion.

Comment (nicolas-grekas) [reply src/Composer/Repository/FilesystemRepository.php]: Here you are updated.

**Comment by nicolas-grekas** [comment]
(rebased)

File: src/Composer/InstalledVersions.php [hunk 5]
```diff
@@ -53,13 +53,14 @@ public static function getInstalledPackages()
      * This also returns true if the package name is provided or replaced by another package
      *
      * @param  string $packageName
+     * @param  bool   $noDevRequirements
      * @return bool
      */
-    public static function isInstalled($packageName)
+    public static function isInstalled($packageName, $noDevRequirements = false)
```
Comment (Seldaek) [inline src/Composer/InstalledVersions.php:59]: Can we instead have a `$includeDevRequirements = true` to avoid the double negation?

Comment (nicolas-grekas) [reply src/Composer/InstalledVersions.php]: sure thing, updated.

**Comment by Seldaek** [comment]
Thanks, merging now that master targets 2.1, but be aware this won't be in a release yet for a while :)
